challenge 2/client-side-application

tomipaul opened the pull request:
Implement client-side application with React and Redux
* Setup webpack
* Implement Actions
* Implement Reducers
* Implement Store
* Implement Components
* Write tests for actions and reducers

tomipaul committed:
[style]: Improve page styles with materialize

tomipaul committed:
[style]: Improve page styles with materialize

tomipaul committed:
[feat] add logo and avatars

tomipaul committed:
[feat] add logo and avatars

tomipaul committed:
[feat] add page for message board

tomipaul committed:
[feat] add page for message board

tomipaul committed:
[chore] update babel configurations

tomipaul committed:
[chore] update babel configurations

tomipaul committed:
[chore] update dependencies and scripts

tomipaul committed:
[chore] update dependencies and scripts

tomipaul committed:
[chore] ignore package-lock.json

tomipaul committed:
[chore] ignore package-lock.json

tomipaul committed:
[feat] update templates for front-end

tomipaul committed:
[feat] update templates for front-end

tomipaul committed:
[chore] update dependencies

tomipaul committed:
[chore] update dependencies

tomipaul committed:
[chore]: update eslint configurations

tomipaul committed:
[chore]: update eslint configurations

tomipaul committed:
[chore]: update project dependencies

tomipaul committed:
[chore]: update project dependencies

tomipaul committed:
[chore] update project dependencies

tomipaul committed:
[chore] update project dependencies

tomipaul committed:
[chore]: update babel configuration

tomipaul committed:
[chore]: update babel configuration

tomipaul committed:
[chore]: update project scripts and dependencies

tomipaul committed:
[chore]: update project scripts and dependencies

tomipaul committed:
[feat]: add action types as string constants

tomipaul committed:
[feat]: add action types as string constants

tomipaul committed:
[feat]: add functions that create action objects

synchronous action creators return plain objects as actions
asynchronous action creators make network calls and call synchronous action creators

tomipaul committed:
[feat]: add functions that create action objects

synchronous action creators return plain objects as actions
asynchronous action creators make network calls and call synchronous action creators

tomipaul committed:
[chore]: add webpack configurations for bundling modules

tomipaul committed:
[chore]: add webpack configurations for bundling modules

tomipaul committed:
[feat]: add css styles to html pages

tomipaul committed:
[feat]: add css styles to html pages

tomipaul committed:
[feat]: add images

tomipaul committed:
[feat]: add images

tomipaul committed:
[feat] add script for signup/login page

tomipaul committed:
[feat] add script for signup/login page

tomipaul committed:
[feat] add landing page

tomipaul committed:
[feat] add landing page

tomipaul committed:
[chore] update configurations

tomipaul committed:
[chore] update configurations

tomipaul committed:
[chore] update reference to bundle output

tomipaul committed:
[chore] update reference to bundle output

tomipaul committed:
[chore] update configurations

change output path in webpack config
ignore bundle output

tomipaul committed:
[chore] update configurations

change output path in webpack config
ignore bundle output

tomipaul committed:
[feat] add entry point for application

tomipaul committed:
[feat] add entry point for application

tomipaul committed:
[feat] add reducers

reducers receive actions and manage state

tomipaul committed:
[feat] add reducers

reducers receive actions and manage state

tomipaul committed:
[feat] create and configure store

create store from combined reducers and apply middlwares for logging and asynchronous actions

tomipaul committed:
[feat] create and configure store

create store from combined reducers and apply middlwares for logging and asynchronous actions

tomipaul committed:
[feat] add presentation and container components

presentation components display views
container components connect presentation components to store to access state and dispatch actions

tomipaul committed:
[feat] add presentation and container components

presentation components display views
container components connect presentation components to store to access state and dispatch actions

tomipaul committed:
[test] ensure action creators create and return appropriate action objects

tomipaul committed:
[test] ensure action creators create and return appropriate action objects

tomipaul committed:
[feat] mock localStorage

tomipaul committed:
[feat] mock localStorage

tomipaul committed:
[test] ensure reducers return appropriate state when given actions

tomipaul committed:
[test] ensure reducers return appropriate state when given actions

tomipaul committed:
[chore] update gitignore

tomipaul committed:
[chore] update gitignore

tomipaul committed:
[chore] modify configurations and asset references for production

tomipaul committed:
[chore] modify configurations and asset references for production

andela-moe commented on client/components/container/Root.jsx:
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Provider } from 'react-redux';
+import { BrowserRouter } from 'react-router-dom';
+import configureStore from '../../store/store';
```
A base filename should exactly match the name of its default export. See more details [here](https://github.com/airbnb/javascript#naming--filename-matches-export)

tomipaul replied:
Okay, I have renamed the file `store.js` to `configureStore.js`

andela-moe commented on client/components/container/Routes.jsx:
```diff
@@ -0,0 +1,43 @@
+        render={() => {
+          return (!token) ?
+          (<Redirect to="/" />) :
+          (<LoadDashboard />);
```
Please use well descriptive names for your components.

tomipaul replied:
@andela-moe thanks for the feedback.
```
Footer has been renamed to MessageArea
Home has been renamed to AuthenticationPage
priorityWasSelected was renamed to selectPriority
messageWasChanged was renamed to changeMessage
```

andela-moe reviewed (commented):
Please address comments and other occurences

andela-moe commented:
Ensure that you resolve your code climate issue and have a good enough GPA for your repo

tomipaul commented on client/components/container/AuthenticateUser.jsx:
```diff
@@ -0,0 +1,21 @@
+  };
+};
+
+const AuthenticateUser = connect(
```
```
connect(null, {singUpUser, SignInuser})(Authentication)
```

tomipaul replied:
I have changed the mapDispatchToProps argument from a function to an object;
`{ SignInUser, SignUpUser }`

tomipaul commented on client/components/presentation/Authentication/SignUp.jsx:
```diff
@@ -0,0 +1,107 @@
+ * @class SignUp
+ * @extends React.Component
+ */
+class SignUp extends React.Component {
```
Convert both signuP and signIn to a functional component

tomipaul replied:
Okay, this has been resolved.

tomipaul commented on client/components/presentation/Common/CommonModal.jsx:
```diff
@@ -0,0 +1,50 @@
+ * @class CommonModal
+ * @extends React.Component
+ */
+class CommonModal extends React.Component {
```
Convert to a functional component

tomipaul replied:
CommonModal component has been removed. it was aimed at DRYing up the code but in the course of refactoring, I saw it did little to achieve that purpose so I let it go.

tomipaul commented on client/components/presentation/Footer/Priority.jsx:
```diff
@@ -0,0 +1,35 @@
+import PropTypes from 'prop-types';
+
+const Priority = ({ priorityType, onClick }) => {
+  $(document).ready(() => {
```
This is bad. change it to use react states to make the update seamless

tomipaul replied:
The functional component has been converted to a class component and the script moved into componentDidMount of the component.

tomipaul commented on client/components/presentation/Footer/PrioritySelect.jsx:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <div className="select-priority-2 col s2">
+      {
+        ['normal', 'urgent', 'critical'].map((priorityType) => {
```
```+        ['normal', 'urgent', 'critical'].map((priorityType) => {
 +          (
 +            <Priority
 +              key={priorityType}
 +              priorityType={priorityType}
 +              onClick={priorityWasSelected}
 +            />
 +          );
 +        })
 +      }
 +    </div>);```

tomipaul replied:
It gives eslint warnings when I remove return statement

tomipaul replied:
This has been resolved. The right configuration was set in the .eslintrc

tomipaul commented on client/components/presentation/Main/Index.jsx:
```diff
@@ -0,0 +1,197 @@
+      searchUserView: false,
+      groupMemberView: false
+    });
+    $('.board-body, footer').removeClass('show-right-bar');
```
This is bad. change it. className

tomipaul commented on client/components/presentation/Main/SideNav/Index.jsx:
```diff
@@ -0,0 +1,62 @@
+const SideNav = ({
+  imageLink,
+  username,
+  groups,
```
groups = {}

tomipaul replied:
Okay, I have effected the change

tomipaul commented on client/components/presentation/Main/SideNav/Index.jsx:
```diff
@@ -0,0 +1,62 @@
+      {
+        (groupIdArray.length !== 0) ?
+        (
+          groupIdArray.map((groupId) => {
```
`Object.keys(groups)`

tomipaul replied:
Okay, I have effected the change

tomipaul commented on client/components/presentation/Modals/UploadPictureModal.jsx:
```diff
@@ -0,0 +1,88 @@
+        }}
+      >Upload<i className="material-icons right">send</i></a>
+    );
+    return super.render();
```
This is super bad!.

tomipaul replied:
Okay, the component has been refactored.

tomipaul commented on client/reducers/activeGroupUsers.js:
```diff
@@ -0,0 +1,36 @@
+      .findIndex((user) => {
+        return user.username === action.response.username;
+      });
+      groupUsers.splice(indexOfRemovedUser, 1);
```
Don't use splice. splice mutate the array. use slice instead.

tomipaul replied:
Code has been refactored. Array.filter method is used to implement logic.

tomipaul commented on client/reducers/userGroups.js:
```diff
@@ -0,0 +1,45 @@
+        groupsObject[group.id] = group;
+        return groupsObject;
+      }, {});
+      const groupsById = action.response.groups
```
```const groupsById = actions.response.groups.map(group => group.id)```

tomipaul replied:
Okay, I have effected the change

tomipaul committed:
[feat] add error dialog to dashboard

tomipaul committed:
[feat] add error dialog to dashboard

tomipaul committed:
[feat] implement review feedback

tomipaul committed:
[feat] implement review feedback

tomipaul committed:
[feat] delete common modal file

tomipaul committed:
[feat] delete common modal file

tomipaul committed:
Merge branch 'review/2' into client

tomipaul committed:
[feat] update MessageArea component

MessageArea keeps state for each group; message input and selected priority are kept until submission
user can navigate to another group without loosing message and priority input

tomipaul committed:
[fix] change default user profile picture and modify profile view

use a silhouette that is not gender specific

tomipaul committed:
[fix] update group members when selected group changes

Before, group members were not updated with every change in selected group
Now, everytime a user selects a group to view, group members are updated

tomipaul committed:
[feat] add logic to get all users of the application

tomipaul committed:
feat(client): implement search user

a registered user can search for another registered user in the application

tomipaul committed:
[fix] allow space characters in group name

tomipaul committed:
[fix] allow space characters in group name

tomipaul committed:
[feat] add tooltips to create group and logout buttons

tomipaul committed:
[feat] add tooltips to create group and logout buttons

tomipaul committed:
[feat] implement markdown

user can write message using the markdown syntax

tomipaul committed:
[feat] implement markdown

user can write message using the markdown syntax

tomipaul committed:
fix merge conflicts

tomipaul committed:
Merge branch 'review/2' of https://github.com/tomipaul/PostIt into review/2

tomipaul committed:
[fix] clear form inputs on form submit

On clicking submit, form inputs should be submitted and cleared

tomipaul committed:
[fix] order messages by time of creation

when getting a group messages, they should be ordered by createdAt

tomipaul committed:
[feat] implement real time messaging using SSE

user should be able to see messages posted to groups in real time

[Delivers #150414713]

tomipaul committed:
[refactor] abstract response handler

a separate middleware takes care of sending response to client

tomipaul committed:
[feat] add email and sms notifications

when a message is posted to a group, members of the group receive notifications based on the priority of the message

tomipaul committed:
Merge feature/150391511/search-registered-users into client

tomipaul committed:
[fix] mount middleware on appropriste path

tomipaul committed:
[feat] add in-app notification

users should see notifications in-app when messages are posted to groups
user must belong to group to receive notifications for the group messages
[Delivers #150391489]

tomipaul committed:
[feat] implement log out

logged in users should be able to log out
[Delivers #150391454]

tomipaul committed:
[feat] track unread messages and acrhive read messages

unread messages are marked as read when a user opens it and read messages are archived
[Delivers 150391587]

tomipaul committed:
[feat] read and archive read messages

messages are marked as read once a user has opened them
Read messages should not clutter a user's message board

tomipaul committed:
[feat] display users that have read a message

Users should be able to see the friends in the groups who have read the messages sent out.
[Delivers #150391503]

tomipaul committed:
[test] fix broken tests

tomipaul committed:
[test] fix broken client-side tests

tomipaul committed:
[test] fix broken client-side tests

tomipaul committed:
[test] ensure reducers are functional

reducers should return a new correct state, given a previous state and action

tomipaul committed:
[test] ensure reducers are functional

reducers should return a new correct state, given a previous state and action

tomipaul committed:
[test] ensure components function appropriately

tomipaul committed:
[test] ensure components function appropriately

tomipaul committed:
[fix] exclude message author from receiving notification

When messages are posted to group, notifications are sent to memebers of the group
Before the author of the message also receives the notification
Now, the author of the message is excluded when sending notifications

tomipaul committed:
[chore] extract styles into a separate css file

During webpack build, every styles chunk should be extracted into a separate css file

tomipaul committed:
[test]: add tests for components

tomipaul committed:
[test]: add tests for components

tomipaul committed:
[refactor] implement review feedback

rename index files so as to import them appropriately
add a 404 error component to be displayed when a non-existing route is navigated to
move network requests from component constructor to componentWillMount
fix button that brings out the sidenav on smaller screens
Add jsdoc comments to all components; class and functional
move logic that process user groups when fetched from reducer into action
rename variables appropriately
change inlign submit functions to class method in SignUp and SignIn components
remove user from search results if user has been added to the active group
add a default message that is shown when a post has not been viewed by any user

tomipaul committed:
[test] add test cases for reducers and actions

tomipaul committed:
Merge branch 'client' into review/3

tomipaul committed:
remove duplicate imports

tomipaul committed:
Merge pull request #21 from tomipaul/review/3

Review/3

tomipaul committed:
Merge pull request #22 from tomipaul/client

150945680 frontend application tests

tomipaul committed:
[tests] fix failing action and reducer test

tomipaul committed:
[tests] fix failing action and reducer test

tomipaul committed:
[test] add component tests

tomipaul committed:
[test] add component tests

tomipaul committed:
[feat]: modify button that add user to a group

add an icon to the button for better user experience

tomipaul committed:
[style] modify style of box that displays group member

tomipaul committed:
[feat] display message for new users

add message that displays on the board of users with no groups

tomipaul committed:
[style] add scss styles

tomipaul committed:
[refactor] modify user experience in adding user to group

tomipaul committed:
[feat] add version to api routes

tomipaul committed:
[fix] add version to routes called in actions

tomipaul committed:
[feat] add reset password feature

add password reset for users on the sign in page
[Delivers #150391435]

tomipaul committed:
[feat] add reset password feature

add password reset for users on the sign in page
[Delivers #150391435]

tomipaul committed:
[test] add version to routes in tests

tomipaul committed:
Merge branch 'chore/150945680/frontend-application-tests' of https://github.com/tomipaul/PostIt into chore/150945680/frontend-application-tests

tomipaul committed:
[chore] add postinstall script

tomipaul committed:
[chore] add p flag to the webpack build command

tomipaul committed:
[chore] add node-sass to dependencies

tomipaul committed:
Merge branch 'feature/150391435/password-reset' of https://github.com/tomipaul/PostIt into feature/150391435/password-reset

tomipaul committed:
[fix] fix travis build

tomipaul committed:
docs: add api documentation

tomipaul committed:
[fix] ignore bin directory when instrumenting files for coverage

tomipaul committed:
[fix] exclude client folder from coverage

tomipaul committed:
[fix] change export to export default

tomipaul committed:
[test] add test for reset Password action

tomipaul committed:
[test] ensure search input is cleared

search input should be cleared when user selects another group

tomipaul committed:
[test] mock react router link component

tomipaul committed:
[fix]: change api version in routes

change v0 to v1 in routes

tomipaul committed:
[fix] change v0 to v1 in mock api routes

tomipaul committed:
[fix]: change v0 to v1 in server routes test
